### PR TITLE
Nullify fs when closing the OpenArchivePathTree

### LINKED ISSUE #41158: Investigate SharedOpenArchivePathTree users counting
I was having a look at a dump provided by @brunobat and I saw quite a lot of:

Screenshot from 2024-06-12 14-20-40

I had a look at the code of `SharedOpenArchivePathTree` here: https://github.com/quarkusio/quarkus/blob/3c6e8f9962e5c2d5c29f52458a10c2abe19afae5/independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/SharedArchivePathTree.java#L68-L110 .

And I found something a bit suspicious as to how the `users` are counted:

- We initialize `users` at `1`
- We increment it in the constructor
- We also increment it on every `acquire()`
- When closing, we check if `decrementAndGet()` is `0`

Let's take a very simple example of me just creating an instance of `SharedOpenArchivePathTree` and closing it right away:

- `users` will be `2` after the constructor call
- when closing and decrementing it, it will be `1` and will not be closed

Now I might have missed something obvious but in that case, a comment in the code would probably be welcome.

### PR DESCRIPTION
**Pull request by gsmet**
When closing a ZipFileSystem, there is a non trivial amount of data that is still kept around and in particular the cen field.

Partially fixes #41158

### CODE REVIEW

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/ArchivePathTree.java [hunk 1]
```diff
@@ -251,7 +252,7 @@ protected void initMultiReleaseMapping(Map<String, String> mrMapping) {
 
         @Override
         public boolean isOpen() {
-            return fs.isOpen();
+            return fs != null && fs.isOpen();
```
Comment (aloubyansky) [inline independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/ArchivePathTree.java:255]: I'd acquire a read lock for this check too

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/ArchivePathTree.java [hunk 2]
```diff
@@ -336,6 +337,10 @@ public void close() throws IOException {
                     throw e;
                 } finally {
                     lock.writeLock().unlock();
+                    // even when we close the fs, everything is kept as is in the fs instance
+                    // and typically the cen, which is quite large
+                    // let's make sure the fs is nullified for it to be garbage collected
+                    fs = null;
```
[the remark this answers is not in the document]
Comment (gsmet) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/ArchivePathTree.java:351]: Done!

File: independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/ArchivePathTree.java (hunk 1 above)
Comment (gsmet) [reply independent-projects/bootstrap/app-model/src/main/java/io/quarkus/paths/ArchivePathTree.java]: Oh, yes, you're right, silly mistake!

**Comment by gsmet** [comment]
@aloubyansky your concerns should be addressed now.

**Review by aloubyansky — APPROVED** [review]

**Pull request merged**